lib/services/graph-progress.js

```diff
@@ -0,0 +1,232 @@
+     */
+    GraphProgressService.prototype.publishGraphStarted = function(graph, swallowError) {
+        return this._publishGraphProgress(graph, 'started', swallowError);
+    };
+
+    /**
+     * Publish a graph finished progress event over AMQP and via Web Hooks.
+     *
+     * @param {Object} graph
+     * @param {String} graphState
+     * @param {Boolean} swallowError
+     * @returns {Promise}
+     * @memberOf GraphProgressService
+     */
+    GraphProgressService.prototype.publishGraphFinished = function(
+        graph,
+        graphState,
+        swallowError
+    ) {
+        return this._publishGraphProgress(graph, graphState, swallowError);
+    };
+
+    /**
+     * Publishes a task started event over AMQP and via Web Hooks
+     *
+     * @param {Object} task
+     * @param {Boolean} swallowError
+     * @returns {Promise}
+     * @memberOf GraphProgressService
+     */
+    GraphProgressService.prototype.publishTaskStarted = function(task, swallowError) {
+        return waterline.graphobjects.findOne({ instanceId: task.context.graphId })
+        .then(function(graph) {
+            var taskFriendlyName = _.get(task, 'definition.friendlyName');
+            var taskId = task.instanceId;
+            var progressMaximum = _.size(_.get(task, 'definition.options.progressMilestones')) ||
+                    Constants.Progress.DefaultTaskProgressMaximum;
+            var taskProgress = {
+                value: 0,
+                maximum: progressMaximum,
+                description: "Task started"
+            };
+
+            var graphDescription = 'Task "' + taskFriendlyName + '" started';
+            var progress = GraphProgress.create(graph, graphDescription);
+            progress.updateTaskProgress(taskId, taskProgress);
+            return eventsProtocol.publishProgressEvent(graph.instanceId,
+                                                       progress.getProgressEventData());
+        })
+        .catch(function(error) {
+            if(swallowError) {
+                logger.error('Error publishing progress event when task started', {
+                    taskId: task.instanceId,
+                    graphId: task.context.graphId,
+                    error: error
+                });
+            } else {
+                throw error;
+            }
+        });
+    };
+
+    /**
+     * Publish a task finished progress event over AMQP and via Web Hooks.
+     *
+     * @param {Object} task
+     * @param {String} task.graphId - the unique ID of the graph to which the task belongs
+     * @param {String} task.taskId - the unique ID of the task
+     * @param {Boolean} swallowError
+     * @returns {Promise}
+     * @memberOf GraphProgressService
+     */
+    GraphProgressService.prototype.publishTaskFinished = function(task, swallowError) {
+        assert.object(task, 'task');
+        assert.uuid(task.graphId, 'task.graphId');
+        assert.uuid(task.taskId, 'task.taskId');
```

Comment:
If the function is designed to return Promise, then all code that potentially throw error need be embedded in Promise, if you don't follow this practice, you may induce the caller a bug or even a process crash!

This is because you never know how your API will be called and you are never sure the input argument is always valid.

For example, if I call your API via following example, seems everything is fine:
```javascript
return doSomething()
.then(function() {
  return publishTaskFinished({graphId: 'xxx', taskId: 'xxx'}, false);
});
```

Then how about below, image the input `task` is invalid and triggered the assertion failure. This function directly throws exception rather than return a rejected Promise, so the latter `catch` in example code actually never catch the error.
```javascript
return publishTaskFinished(null, false)
  .then(function() { doSomethingElse() })
  .catch(function(err) {
     console.log(err);
  });
```

With similar code as above, I directly call the function `f()`, if the input `task` is invalid, the `f()` will throw error, and the error is never caught, so the whole process crash!!
```javascript
function f(task) {
  return publishTaskFinished(task, false)
    .then(function() { doSomethingElse() });
}

f(null);
```